packages/demo/frontend/src/components/EarnWithFrontendWallet.tsx

```diff
@@ -0,0 +1,474 @@
+      setTimeout(async () => {
+        if (wallet && marketData) {
+          try {
+            await fetchPosition()
+          } catch {
+            setDepositedAmount('0.00')
+          }
+        }
+      }, 1000)
+
+      // Also refresh wallet balance
+      if (wallet) {
+        setTimeout(async () => {
+          await fetchBalance()
+        }, 2000)
+      }
+
+      return {
+        transactionHash: txHash,
+        blockExplorerUrl: explorerUrl,
+      }
+    },
+    [wallet, marketData, fetchBalance],
+  )
+
+  if (!isLoggedIn) {
+    return <LoginWithDynamic />
+  }
+
+  return (
+    <Earn
+      ready={true}
+      selectedProvider={selectedProvider}
+      walletAddress={wallet?.address || null}
+      logout={logout}
+      usdcBalance={usdcBalance}
+      isLoadingBalance={isLoadingBalance}
+      apy={apy}
+      isLoadingApy={isLoadingApy}
+      depositedAmount={depositedAmount}
+      isLoadingPosition={isLoadingPosition}
+      isInitialLoad={isInitialLoad}
+      onMintUSDC={handleMintUSDC}
+      onTransaction={handleTransaction}
+    />
+  )
+}
+
+async function formatMarketResponse(market: LendMarket) {
+  return {
+    marketId: market.marketId,
+    name: market.name,
+    asset: market.asset,
+    supply: {
+      totalAssets: formatUnits(
+        market.supply.totalAssets,
+        market.asset.metadata.decimals,
+      ),
+      totalShares: formatUnits(market.supply.totalShares, 18),
+    },
+    apy: market.apy,
+    metadata: market.metadata,
+  }
+}
+
+async function formatMarketBalanceResponse(
+  balance: LendMarketPosition,
+): Promise<{
+  balance: string
+  balanceFormatted: string
+  shares: string
+  sharesFormatted: string
+}> {
+  return {
+    balance: balance.balanceFormatted,
+    balanceFormatted: balance.balanceFormatted,
+    shares: balance.sharesFormatted,
+    sharesFormatted: balance.sharesFormatted,
+  }
+}
+
+function isEOATransactionReceipt(
+  receipt: LendTransactionReceipt,
+): receipt is EOATransactionReceipt {
+  return !Array.isArray(receipt) && !('userOpHash' in receipt)
+}
+
+function isBatchEOATransactionReceipt(
+  receipt: LendTransactionReceipt,
+): receipt is EOATransactionReceipt[] {
+  return Array.isArray(receipt)
+}
+
+function isUserOperationTransactionReceipt(
+  receipt: LendTransactionReceipt,
+): receipt is UserOperationTransactionReceipt {
+  return 'userOpHash' in receipt
+}
+
+async function getBlockExplorerUrls(
+  chainId: SupportedChainId,
+  transactionHashes?: string[],
+  userOpHash?: string,
+): Promise<string[]> {
+  const chain = chainById[chainId]
+  if (!chain) {
+    throw new Error(`Chain not found for chainId: ${chainId}`)
+  }
+
+  let url = `${chain.blockExplorers?.default.url}`
+  if (chain.id === unichain.id) {
+    url = `https://unichain.blockscout.com`
+  }
+  if (chain.id === baseSepolia.id) {
+    url = `https://base-sepolia.blockscout.com`
+  }
+
+  if (userOpHash) {
+    return [`${url}/op/${userOpHash}`]
+  }
```

Comment:
these were copy/pasted from the backend service. In a follow up pr I plan to deduplicate this by removing formatting from the backend and consolidating all formatting to the frontend app.